text/0000-project-error-handling.md

```diff
@@ -0,0 +1,121 @@
+
+# Charter
+[charter]: #charter
+
+## Goals
+
+### Agree on and define common error handling terminology
+
+- Recoverable error: An error that can reasonably be expected to be encountered e.g. a missing file.
+- Unrecoverable error: An error that cannot reasonably be expected to happen and which indicates a bug e.g. indexing out of bounds.
+- Error Type: A type that represents a recoverable error. Error types can optionally implement the error trait so that it can be reported to the user or be converted into a trait object.
+- Reporting Type: A type that can store all recoverable errors an application may need to propogate and print them as error reports.
+    - Reporting types can represent the recoverable errors either via concrete types, likely paramaterized, or trait objects.
+    - Reporting types often bundle context with errors when they are constructed, e.g. `Backtrace`.
+    - Reporting types often provide helper functions for creating ad hoc errors whose only purpose is to be reported e.g. `anyhow::format_err!` or `eyre::WrapErr`.
+
+### Come to a consensus on current best practices
+
+Here is a tenative starting point, subject to change:
+
+- Use `Result` and `Error` types for recoverable errors.
+- Use `panic` for unrecoverable errors.
+- Impl `Error` for error types that may need to be reported to a human or be composed with other errors.
+- Use enums for types representing multiple failure cases that may need to be handled.
+    - For libraries, oftentimes you want to support both reporting and handling so you impl `Error` on a possibly non-exhaustive enum.
+- Error kind pattern for associating context with every enum variant without including the member in every enum variant.
+- Convert to a reporting type when the error is no longer expected to be handled beyond reporting e.g. `anyhow::Error` or `eyre::Report` or when trait object + downcast error handling is preferable.
+- Recommend `Box`ing concrete error types when stack size is an issue rather than `Box`ing and converting to `dyn Error`s.
+- What is the consensus on handling `dyn Error`s? Should it be encouraged or discouraged? Should we look into making `Box<dyn Error...>` impl Error?
+
+
+### Communicate current best practices
+
+- Document the concensus.
+- Communicate plan for future changes to error handling, and the libraries that future changes are being based off of.
+- Produce learning resources related to current best practices.
+    - New chapters in the book?
+
+### Evaluate options for error reporting type aka better `Box<dyn Error>`
+
+- Survey the current libraries in the ecosystem:
+    - `anyhow`
+    - `eyre`
+- Evaluate value of features including:
+    - Single word width on stack
+    - Error wrapping with display types and with special downcast support.
+    - report hook + configurable `dyn ReportHandler` type for custom report formats and content, similar to panic handler but for errors.
+    - libcore compatibility.
```

Comment:
Everything should be compatible with libcore. I think you want to phrase this as "doesn't require libstd"

Reply:
how about

```suggestion
    - `#[no_std]` support
```